Added lookup_with_etc method

tgrave opened the pull request:
Added lookup_with_etc method to get timezones for the whole Earth surface (including oceans and other "Etcetera" areas). Geonames only (since Google API returns ZERO_RESULTS for Etc areas).

panthomakos commented on lib/timezone/lookup/geonames.rb:
```diff
@@ -34,13 +34,37 @@ def lookup(lat, long)
 
+      def get_etc(offset)
+        return 'Etc/UTC' if offset == 0
+        "Etc/GMT#{format('%+d', offset.to_i)}"
```
This seems like good default behavior. I don't see why it only needs to be implemented with Geonames?

Even if it is only implemented with geonames, why should this not be included as part of the default `lookup` function?

tgrave replied:
Well, now for coordinates in the ocean, for example, lookup will rise exception, because both geonames and google do not give timezoneId for such points.

I thought that about including it in the lookup function, but I was not sure about compatibility issues. After all someone may use the fact thel lookup(0,0) raises exception, and when method's behaviour changes, it may couse some problems for people who use this gem.

Maybe it would be better to use it as an option for lookup, like this:
`
def lookup(lat, long, options = {})
...
  return get_etc(data['gmtOffset']) if options[:with_etc] && data['gmtOffset']
...
end  
`

I'll refactor it to include into lookup method, but should I do it optional to not change current behaviour for compatibility?

panthomakos replied:
You don't need to worry about compatibility in this case. `lookup(0,0)` may raise an exception today and work tomorrow even if there is no gem change - simply because the API results change. The API is still the same, so you aren't breaking backwards compatibility.

Thinking about this further, my only concern is that post-processing may be confusing because this code is trying to make a best guess for what a timezone would be if a land-form existed there. For example, while it is reasonable to expect that the middle of the Pacific Ocean has an `Etc/UTC+/-..` timezone, it is also not really correct because there is no official timezone there - because there is no land there.

I would vote for adding a configuration option to `Timezone::Lookup.config(:geonames)` like:

```
Timezone::Lookup.config(:geonames) do |c|
  c.offset_etc_zones = true
end
```

The default value would be `false`.

You can then use that configuration option in the `#lookup` function to include or exclude responses w/o a timezone id.

Does this make sense?

tgrave replied:
Yes, it makes sense.

Though, I don't think it's really guessing. Well, mostly not guessing. Guessing would be if gmtOffset depended on longitude only, but Geonames API returns gmtOffset with consideration of the real timezones borders. For example, let's take the the Pacific Ocean, for lat=20,long=165 it returns gmtOffset=12, but for lat=-20,long=165 it returns gmtOffset=11. Even in the ocean the borders of timezones are not straight, in this case we can see timezone change in a huge mostly empty area Marshall Islands.

On the other hand there are certain areas where offset is fractional, i.e. for the sea near Norfolk Island with lat=-29,long=167 returned gmtOffset is 11.5. And for such rare cases it would be really guessing as there are no Etc timezones with fractional offset. I just truncate such offset to the integer. In fact for Norfolk Island it would be the right thing to do since they changed their offset to 11 last October (Geonames just does not know about it obviously).

But anyway, I agree with making it optional for lookup method and putting it in the configuration.

panthomakos commented on test/timezone/lookup/test_geonames.rb:
```diff
@@ -35,6 +49,19 @@ def test_lookup
   end
 
+  def test_lookup_with_etc
+    %w(arctic atlantic).each do |point|
```
If you convert the data into a more structured hash this can be simplified.

```
def etc_data
  {
    arctic: { coordinates: [89, 40], name: 'Etc/GMT+3' },
    atlantic: {... }
  }
end

def test_lookup_with_etec
  etc_data.each do |key, data|
    ...
    assert_equal data[:name], mine.lookup(*data[:coordinates])
```

tgrave replied:
Yes, you are right of course. I just copypasted test from the lookup method in rough-and-ready fashion, so did not made a proper changes for multiple points test. I'll change it.

panthomakos commented on lib/timezone/lookup/geonames.rb:
```diff
@@ -36,11 +34,21 @@ def lookup(lat, long)
+
+        if config.offset_etc_zones && data['gmtOffset']
+          return 'Etc/UTC' if data['gmtOffset'] == 0
+          "Etc/GMT#{format('%+d', data['gmtOffset'].to_i)}"
```
Why do you `#to_i` here? This should already be a number, right?

```
'%+d' % 11.5
=> "+11"
```

tgrave replied:
Oh, it's just a safeguard in case API returns something other than a number. But maybe ArgumentError exception would be better if API returns unexpected data.

panthomakos replied:
I don't think returning an `ArgumentError` is good because the programmer didn't provide a bad argument and this approach would definitely break from the existing API.

At the least we should return `nil` instead of trying to coerce the correct value - as this will have the same effect as having an invalid timezone. If the value is not a Numeric, the integer value will be `0` - which we don't really want.

Maybe something like this?

```
return unless data['gmtOffset'].is_a?(Numeric)
rest of your code as is
```

A test for this scenario would also be good so that the behavior is asserted.

tgrave replied:
Yes, I've added nil resilt if offset is not numeric, and a test for this case too.

panthomakos commented on test/timezone/lookup/test_geonames.rb:
```diff
@@ -9,10 +9,19 @@ def coordinates
+    }
+  end
+
+  def lookup(options = {})
```
Maybe we can just yield here to make additional option setting easier:

```
def lookup(&block)
  ...
  config.request_handler = HTTPTestClient
  yield(config) if block_given?
  
  Timezone::Lookup::Geonames.new(config)
end
```

then

```
mine = lookup { |c| c.offset_etc_zones = trues }
```

panthomakos commented:
Looking good. Can you add a CHANGELOG entry and some README information about the new option?

panthomakos commented:
P.S: RuboCop failures have been fixed in master.

tgrave commented:
Ok. Added a CHANGELOG entry and a README section. Also fixed my own mistake with offset sign for Etc/GMT zones - added a note on this at the end of the README section.

panthomakos commented:
I'm also fine with the simple to_i approach as long as the behavior is
tested and the results make sense.

On Sun, May 29, 2016, 07:53 Sergey Tsvetkov <notifications@github.com>
wrote:

> Ok. Added a CHANGELOG entry and a README section. Also fixed my own
> mistake with offset sign for Etc/GMT zones - added a note on this at the
> end of the README section.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/panthomakos/timezone/pull/62#issuecomment-222364784>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AALYOLg272IrDrBxLcrCnS3NqTgCXQFDks5qGahqgaJpZM4IlaH8>
> .
>

panthomakos commented on lib/timezone/lookup/geonames.rb:
```diff
@@ -25,9 +25,7 @@ def lookup(lat, long)
-        return data['timezoneId'] if data['timezoneId']
-
-        return unless data['status']
+        return get_timezone_id(data) unless data['status']
```
One final comment. This is a bit of a behavior change. I don't want to end up in the case where a `timezoneId` and `status` is returned and we fail to return.

```
timezoneId: 'America/Los_Angeles',
stauts: 'success'
```

Can we just inline some of the functionality to revert this?

```
return data['timezoneId'] if data['timezoneId']

if data['gmtOffset'] && config.offset_etc_zones
  return calculate_gmt_offset(data['gmtOffset'])
end

return unless data['status']

...
```

tgrave replied:
Inlining it won't work with rubocop - too mych complexity in the lookup method. That's the reason I had moved so much into get_timezone_id method. But I've returned old behaviour with data['status'].

panthomakos commented:
One comment, otherwise LGTM. If you can squash your commits, that would be great, if not I can handle it.

Thank you for the contribution!

panthomakos commented:
Any updates on finishing this up?

tgrave commented:
Sorry, I have fixed old behaviour issue and will try to squash the commits now...

panthomakos commented:
Looks great! Thanks for the contribution!
